tools/system_libs.py

```diff
@@ -286,325 +599,392 @@ def create_libc(libname):
-    # it includes the builtin symbols that the LTO complication can generate.
-    # It seems that LTO as implemented by lld assumes that builtins do not
-    # take part in LTO.
-    # TODO(sbc): If we ever fix https://bugs.llvm.org/show_bug.cgi?id=41384 then
-    # this restriction can be removed.
-    o_s = []
-    commands = []
-    for src in files:
-      o = in_temp(os.path.basename(src) + '.o')
-      commands.append([shared.PYTHON, shared.EMCC, '-fno-builtin', '-O2',
-                       '-c', shared.path_from_root('system', 'lib', src),
-                       '-o', o] +
-                      musl_internal_includes() + extra_flags +
-                      get_cflags(force_object_files=True))
-      o_s.append(o)
-    run_commands(commands)
-    lib = in_temp(libname)
-    shared.Building.emar('cr', lib, o_s)
-    return lib
-
-  def create_wasm_compiler_rt(libname):
-    files = files_in_path(
-      path_components=['system', 'lib', 'compiler-rt', 'lib', 'builtins'],
-      filenames=['addtf3.c', 'ashlti3.c', 'ashrti3.c', 'atomic.c', 'comparetf2.c',
-                 'divtf3.c', 'divti3.c', 'udivmodti4.c',
-                 'extenddftf2.c', 'extendsftf2.c',
-                 'fixdfti.c', 'fixsfti.c', 'fixtfdi.c', 'fixtfsi.c', 'fixtfti.c',
-                 'fixunsdfti.c', 'fixunssfti.c', 'fixunstfdi.c', 'fixunstfsi.c', 'fixunstfti.c',
-                 'floatditf.c', 'floatsitf.c', 'floattidf.c', 'floattisf.c',
-                 'floatunditf.c', 'floatunsitf.c', 'floatuntidf.c', 'floatuntisf.c', 'lshrti3.c',
-                 'modti3.c', 'multc3.c', 'multf3.c', 'multi3.c', 'subtf3.c', 'udivti3.c', 'umodti3.c', 'ashrdi3.c',
-                 'ashldi3.c', 'fixdfdi.c', 'floatdidf.c', 'lshrdi3.c', 'moddi3.c',
-                 'trunctfdf2.c', 'trunctfsf2.c', 'umoddi3.c', 'fixunsdfdi.c', 'muldi3.c',
-                 'divdi3.c', 'divmoddi4.c', 'udivdi3.c', 'udivmoddi4.c'])
-    files += files_in_path(path_components=['system', 'lib', 'compiler-rt'],
-                           filenames=['extras.c'])
-    return create_wasm_rt_lib(libname, files)
-
-  def create_wasm_ubsan_minimal_rt(libname):
-    files = files_in_path(
-      path_components=['system', 'lib', 'compiler-rt', 'lib', 'ubsan_minimal'],
-      filenames=['ubsan_minimal_handlers.cpp'])
-    return create_wasm_rt_lib(libname, files)
-
-  def create_wasm_common_san_rt(libname):
-    files = glob_in_path(
-      path_components=['system', 'lib', 'compiler-rt', 'lib', 'sanitizer_common'],
-      glob_pattern='*.cc',
-      excludes=['sanitizer_common_nolibc.cc'],
+    if self.is_tracing:
+      cflags += ['--tracing']
+    return cflags
+
+  def get_base_name(self):
+    base = 'lib%s' % self.malloc
+
+    extra = ''
+    if self.is_debug:
+      extra += '_debug'
+    if not self.use_errno:
+      # emmalloc doesn't actually use errno, but it's easier to build it again
+      extra += '_noerrno'
+    if self.is_mt:
+      extra += '_threadsafe'
```

Comment:
doesn't inheriting from `MTLibrary` handle adding the `-mt` suffix for us?

if this is overriding just to keep the same names as before, i think it's ok to change the name if it simplifies the code (which it would if just one place adds `-mt` or such to the name).

Reply:
Somewhat unfortunately, due to the way inheritance work, we cannot actually call `super` if we want use the prefix `libdlmalloc` or `libemmalloc`, which is the reason why we are handling it here manually.

Reply:
i probably don't understand yet. why doesn't this problem happen with all the other uses of MTLibrary?

Reply:
All other uses of `MTLibrary` are fine because they end up calling `Library`'s `get_base_name`, returning the `name` class attribute. However, the `name` attribute on `libmalloc` is `libmalloc`, so using inheritance, we'd be forced to have `libmalloc-mt-dlmalloc...`, which looks terrible.

Reply:
I see, thanks.

Reply:
Thinking on this some, I'm not sure malloc is that special, and this issue may happen more over time. If inheritance is clumsy here, perhaps an option is to avoid it? That is, instead of
```
class SomeLibrary(MTLibrary, OtherLibrary):
  ..
```
there could be a single class for all library types, and "Modifiers" (VariantGenerators? Variators?) that are added to a list,
```
class SomeLibrary(Library):
  modifiers = [MTLibrary, OtherLibrary]
  ..
```
then the "Modifiers" would be called in the proper places?

Reply:
Actually now that I think about it, we can still do this with inheritance. For this use case, we just need a method that returns `self.name` that can be overridden. I suppose we might as well do this for `libmalloc` as an example for the future.

Reply:
IMO using multiple inheritance makes the implementation harder to follow but makes the resulting client code more elegant. The code is documented well enough that I don't think the complexity of the implementation will be a maintenance issue going forward, so I'm happy with the current design.

@kripken I agree that your proposed solution would have a simpler implementation, so if you feel strongly about this then I would be happy going in that direction as well.